plugin/src/main/java/org/opensearch/ml/action/agents/DeleteAgentTransportAction.java

```diff
@@ -39,75 +52,123 @@
-                                .onFailure(
-                                    new OpenSearchStatusException(
-                                        "User doesn't have privilege to perform this operation on this hidden agent",
-                                        RestStatus.FORBIDDEN
-                                    )
-                                );
+
+            sdkClient
+                .getDataObjectAsync(getDataObjectRequest, client.threadPool().executor(GENERAL_THREAD_POOL))
+                .whenComplete((r, throwable) -> {
+                    context.restore();
+                    log.debug("Completed Get Agent Request, Agent id:{}", agentId);
+                    if (throwable != null) {
+                        Exception cause = SdkClientUtils.unwrapAndConvertToException(throwable);
+                        if (cause instanceof IndexNotFoundException) {
+                            log.error("Failed to get Agent index", cause);
+                            actionListener.onFailure(new OpenSearchStatusException("Failed to get agent index", RestStatus.NOT_FOUND));
                         } else {
-                            // If the agent is not hidden or if the user is a super admin, proceed with deletion
-                            DeleteRequest deleteRequest = new DeleteRequest(ML_AGENT_INDEX, agentId);
-                            client.delete(deleteRequest, ActionListener.wrap(deleteResponse -> {
-                                log.debug("Completed Delete Agent Request, agent id:{} deleted", agentId);
-                                actionListener.onResponse(deleteResponse);
-                            }, deleteException -> {
-                                log.error("Failed to delete ML Agent " + agentId, deleteException);
-                                actionListener.onFailure(deleteException);
-                            }));
+                            log.error("Failed to get ML Agent {}", agentId, cause);
+                            actionListener.onFailure(cause);
+                        }
+                    } else {
+                        try {
+                            GetResponse gr = GetResponse.fromXContent(r.parser());
+                            if (gr.isExists()) {
+                                try (
+                                    XContentParser parser = jsonXContent
+                                        .createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, gr.getSourceAsString())
+                                ) {
+                                    ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+                                    MLAgent mlAgent = MLAgent.parse(parser);
+                                    if (TenantAwareHelper
+                                        .validateTenantResource(mlFeatureEnabledSetting, tenantId, mlAgent.getTenantId(), actionListener)) {
```

Comment:
`validateTenantResource` method responsibility looks not clear, it should validate tenant , but actually it also covers send back the validation failure through action listener. 
It looks ok for now, but such combined/unclear responsibility make it hard to use this method if there is not actionlistener.

Reply:
It also add some effort for code review. When I review this part, I doubt why there is no `else` block matching this `if`. So I have to read the code of `validateTenantResource` method to learn why. This impacts code readability.

Reply:
@dbwiddis had also similar comment [here](https://github.com/opensearch-project/ml-commons/pull/2588#discussion_r1649758398).  Do you think renaming the method to `validateTenantIdandFailListenerIfInvalid` can help?

In ml-commons we have similar methods like this : [validateModelGroupAccess](https://github.com/opensearch-project/ml-commons/blob/f61ab171f7e36eee727b678ca0f26f46204ae891/plugin/src/main/java/org/opensearch/ml/helper/ModelAccessControlHelper.java#L77C17-L77C41), [validateConnectorAccess](https://github.com/opensearch-project/ml-commons/blob/f61ab171f7e36eee727b678ca0f26f46204ae891/plugin/src/main/java/org/opensearch/ml/helper/ConnectorAccessControlHelper.java#L62C17-L78)

Reply:
Ok, I'll address this in another PR then. I have to touch multiple files in that case.